File: java/src/main/java/com/dsa/examples/sorting/SelectionSort.java
// Sorting — Selection Sort

package com.dsa.examples.sorting;

/**
 * Sort an array in ascending order in place.
 * Technique: Selection sort
 * Invariant: Prefix [0..i) is sorted and contains the i smallest elements.
 * sort: Time O(n²); Space O(1).
 */
public final class SelectionSort {

    private SelectionSort() {
    }

    /**
     * Sorts {@code arr} in ascending order in place.
     * Time: O(n²); Space: O(1).
     */
    public static void sort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return;
        }
        for (int i = 0; i < arr.length - 1; i++) {
            int minIdx = i;
            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[minIdx]) {
                    minIdx = j;
                }
            }
            if (minIdx != i) {
                int t = arr[i];
                arr[i] = arr[minIdx];
                arr[minIdx] = t;
            }
        }
    }
}

File: python/dsa/sorting/selection_sort.py
# Sorting — Selection Sort

"""Sort an array in ascending order in place.

Technique: Selection sort
Invariant: Prefix [0..i) is sorted and contains the i smallest elements.
sort: Time O(n²); Space O(1).
"""


def sort(arr: list[int] | None) -> None:
    """Sorts arr in ascending order in place.

    Time: O(n²); Space: O(1).
    """
    if arr is None or len(arr) < 2:
        return
    for i in range(len(arr) - 1):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        if min_idx != i:
            arr[i], arr[min_idx] = arr[min_idx], arr[i]
